File: src/initFun.c
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include "harlem.h"



void initT(double *a_T1,double *a_T2,double *a_Stimes, double *a_Sjumps,double *a_dGn, double *a_delta, int *a_n,int *a_m, double *a_ww,  double *a_xGrid, double *a_vGrid, int *a_nx, int *a_nv)
{
    for(int i=0; i<*a_nv; i++)
    {
        for(int j=0; j<*a_nx; j++)
        {
	    a_T1[i + *a_nv * j] = 0.0;
	    a_T2[i + *a_nv * j] = 0.0;

            if(a_vGrid[i] > a_xGrid[j] + *a_delta)
            {
                double denominatore = 0.0;
                int k = 0;
		
                while((a_vGrid[i] > a_ww[k]) && (k < *a_n))
                {
                    if(a_ww[k] > a_xGrid[j]) denominatore += a_dGn[k] * kaplanM(a_vGrid[i] - a_ww[k], a_Stimes, a_Sjumps, *a_m);
                    k++;
                }
                if(denominatore > sqrt(DBL_EPSILON))
                {
                    a_T1[i + *a_nv * j] = (a_vGrid[i] - a_xGrid[j])/denominatore;
                    a_T2[i + *a_nv * j] = 1/denominatore;
                }
            } 
        }
    }
}

double kaplanM(double a_val, double *a_Stimes, double *a_Sjumps, int a_m)
{
    if(a_val<a_Stimes[0]) return(1.0);
    if(a_val>a_Stimes[a_m-1]) return(a_Sjumps[a_m-1]);
    
    int up = a_m - 1;
    int low = 0;
    int i = ceil((up+low)/2);
    while(1)
    {
        
        if(a_val == a_Stimes[i+1]) return(a_Sjumps[i]);
        if(a_val == a_Stimes[i]) return(a_Sjumps[i-1]);
        if(a_val == a_Stimes[i-1]) return(a_Sjumps[i-1]);
        
        if(a_Stimes[i]>a_val)
        {
            if(a_Stimes[i-1] <= a_val)
            {
                return(a_Sjumps[i-1]);
            }else
            {
                up = i;
                i = floor((up+low)/2);
            }
        }else if(a_Stimes[i+1]>a_val)
        {
            return(a_Sjumps[i]);
        }else
        {
            low = i;
            i = ceil((up+low)/2);
        }
        
    }
}



//void initQslow(double *a_delta,double *a_ab, double *a_cd, int *a_n, double *a_xx, double *a_vv, double *a_h1, double *a_h2, int *a_n1, int *a_n2,int *a_N, double *a_xGrid, double *a_vGrid, int *a_nx, int *a_nv, double *a_Q2,double *a_hOpt, double *a_Weights)
void initQ(double *a_delta,double *a_tau, int *a_n, double *a_xx, double *a_vv, double *a_h1, double *a_h2, int *a_n1, int *a_n2,int *a_N, double *a_xGrid, double *a_vGrid, int *a_nx, int *a_nv, double *a_Q2,double *a_hOpt, double *a_Weights)
{

    double ISE = 0.0;
    double normC = 0.0;

    if(*a_n1>1 || *a_n2>1)
    {
      //kdeISEslow(a_delta, a_ab, a_cd, a_n, a_xx, a_vv, a_h1, a_h2, a_n1, a_n2, a_hOpt, a_N, a_Weights);
      kdeISE(a_delta, a_tau, a_n, a_xx, a_vv, a_h1, a_h2, a_n1, a_n2, a_hOpt, &ISE, &normC, a_N, a_Weights);
    } else {
	a_hOpt[0] = *a_h1;
	a_hOpt[1] = *a_h2;
    }
    for(int i=0; i<*a_nv;i++)
    {
        for(int j=0;j<*a_nx;j++)
        {
	  /*
	    if(i==0 || j==0)
	    {
		a_Q2[i + *a_nv * j] = 0.0;
		continue;
	    }
	    double y = (a_vGrid[i] - a_xGrid[j] - *a_delta);
	    if(y <= 0)
	    {
		a_Q2[i + *a_nv * j]=0.0;
		continue;
	    }
	  */
	    //a_Q2[i + *a_nv * j] = kde_slow(*a_delta, y, a_xGrid[j], *a_n, a_xx, a_vv, a_hOpt) / (a_xGrid[j] * y);
	    a_Q2[i + *a_nv * j] = kde(*a_delta, *a_tau,a_vGrid[i] - a_xGrid[j], a_xGrid[j], *a_n, a_xx, a_vv, a_hOpt)/normC;
        }
    }
}

void kdeISE(double *a_delta,double *a_tau, int *a_n, double *a_xx, double *a_vv, double *a_h1, double *a_h2, int *a_n1, int *a_n2,double *a_hOpt,double *a_ISE, double *a_Norm,int *a_N, double *a_Weights)
{
    
    double a = 0.0;
    double b = *a_tau - *a_delta;
    double c = *a_delta;
    double d = *a_tau;
    double step = (*a_tau - *a_delta)/((double)*a_N);

    
    double ISE=0.0;
    
    for (int i=0; i<*a_n1; i++)
    {
        
        double h[2];
        h[0]=a_h1[i];
        
        for (int j=0; j<*a_n2; j++)
        {
            
            double avgSq = 0.0;
            double normC = 0.0;
            
            h[1]=a_h2[j];
            
            /// Simpson Formula in 2D
	    for(int l=0;l<=*a_N;l++)
            {
		for(int k=0;k<=*a_N;k++)
                {

		    double term1 = kde(*a_delta,*a_tau,c+l*step,a+k*step,*a_n,a_xx,a_vv, h);
                
		    normC += a_Weights[l+(*a_N+1)*k]*term1;
		    avgSq += a_Weights[l+(*a_N+1)*k]*term1*term1;
                
		}
            }
            normC *= step*step/9.0;
            avgSq *= step*step/9.0;
            avgSq/=normC*normC;
            
            double loo=kdeLoo(*a_delta,*a_tau,*a_n,a_xx,a_vv,h);
            
            loo /= normC;
            
            ISE = avgSq-2*loo;

	    
            if(i==0 && j==0)
            {
                *a_ISE=ISE;
                a_hOpt[0]=h[0]; a_hOpt[1]=h[1];
                *a_Norm=normC;
            }else if (ISE<*a_ISE) 
            {
                *a_ISE=ISE;
                a_hOpt[0]=h[0]; a_hOpt[1]=h[1];
                *a_Norm=normC;
            }
            
        }
    }
    
}


double kde(double a_delta,double a_tau, double a_v, double a_x, int a_n, double *a_xx, double *a_vv, double a_h[2])
{
  
  double y = a_v;
  
  if((y<a_delta)||(y>a_tau)) return(0);
  if((a_x<0)||(a_x>a_tau-a_delta)) return(0);
  
  double p = 0.0;
  
  for (int i=0; i<a_n; i++) 
  {
   
    double yy = a_vv[i] - a_xx[i];
    double t = (y-yy)/a_h[0];
    double z = (a_x-a_xx[i])/a_h[1];

    if((t<1.0)&&(t>-1.0)&&(z<1.0)&&(z>-1.0)) p+=(1-(t*t))*(1-(z*z)); 

  }

  p /= a_n*(a_h[0]*a_h[1])*16/9;
  
  return(p);
	
}


double kdeLoo(double a_delta,double a_tau, int a_n, double *a_xx, double *a_vv, double a_h[2])
{
  
  double LOO=0.0;
  
  for (int i=0; i<a_n; i++) 
  {
  
    double y=a_vv[i]-a_xx[i];
    double x=a_xx[i];
    
    if((y<a_delta)||(y>a_tau)) continue;
    if((x<0.0)||(x>a_tau-a_delta)) continue;
    
    for (int j=0; j<a_n; j++) 
    {
      
      if(i==j)continue;

      double yy=a_vv[j]-a_xx[j];

      if((yy<a_delta)||(yy>a_tau)) continue;
      if((a_xx[j]<0.0)||(a_xx[j]>a_tau-a_delta)) continue;

      double t=(y-yy)/a_h[0];
      double z=(x-a_xx[j])/a_h[1];

      if((t<1.0)&&(t>-1.0)&&(z<1.0)&&(z>-1.0)) LOO+=(1-(t*t))*(1-(z*z));

    }

  }
	
  LOO/=a_n*(a_n-1)*(a_h[0]*a_h[1])*16/9;

  return(LOO);
  	
}



void kdeISEslow(double *a_delta, double *a_ab, double *a_cd, int *a_n, double *a_xx, double *a_vv, double *a_h1, double *a_h2, int *a_n1, int *a_n2,double *a_hOpt, int *a_N, double *a_Weights)
{
    
    double ISE = 0.0;
    double min_ISE = 0.0;

    
    for (int i=0; i<*a_n1; i++)
    {
        
        double h[2];
        h[1]=a_h2[i];
	double c = a_cd[0] - h[1];
	double d = a_cd[1] + h[1];
	double stepU2 = (d - c)/((double)*a_N);
        
        for (int j=0; j<*a_n2; j++)
        {
            
            double avgSq = 0.0;
            h[0]=a_h1[j];

	    //printf("%lf %lf\n",h[0],h[1]);

	    double a = a_ab[0] - h[0];
	    double b = a_ab[1] + h[0];
	    double stepU1 = (b - a)/((double)*a_N);
            
            /// Simpson Formula in 2D
	    for(int k=0;k<=*a_N;k++)
	      {
		for(int l=0;l<=*a_N;l++)
		  {
		    
		    double term1 = kde_slow(*a_delta,c+l*stepU2,a+k*stepU1,*a_n,a_xx,a_vv, h);
		    //double term1 = kde(*a_delta,*a_tau,c+l*step-(a+k*step)- *a_delta,a+k*step,*a_n,a_xx,a_vv, h);
		    avgSq += a_Weights[l+(*a_N+1)*k]*term1*term1;
		    //kde(double a_delta,double a_tau, double a_v, double a_x, int a_n, double *a_xx, double *a_vv, double a_h[2])
		}
            }

            avgSq *= stepU1*stepU2/9.0;
            
            double loo = kdeLoo_slow(*a_delta,*a_n,a_xx,a_vv,h);
                        
            ISE = avgSq-2*loo;
	    
            if(i==0 && j==0)
            {
                min_ISE = ISE;
                a_hOpt[0]=h[0]; a_hOpt[1]=h[1];
            }else if (ISE < min_ISE) 
            {
                min_ISE = ISE;
                a_hOpt[0]=h[0]; a_hOpt[1]=h[1];
            }
            
        }
    }
}


double kde_slow(double a_delta, double a_v, double a_x, int a_n, double *a_xx, double *a_vv, double a_h[2])
{

  if(a_v <= 0.0) return(0.0);
  if(a_x <= 0.0) return(0.0);
  double x = log(a_x); double y = log(a_v);
  
  double p = 0.0;
  
  for (int i=0; i<a_n; i++) 
  {
   
    double yy = a_vv[i] - a_xx[i] - a_delta;

    if(yy <= 0.0) continue;
    if(a_xx[i] <= 0.0) continue;

    double z = (x - log(a_xx[i]))/a_h[0];
    double t = (y - log(yy))/a_h[1];
    //double zz = z*z;
    //double tt = t*t;
    double uu = z*z + t*t;

    //if((tt<1.0) && (tt>-1.0) && (zz<1.0) && (zz>-1.0)) p += (1-(t*t)) * (1-(z*z)); 
    if((uu<1.0) && (uu>-1.0)) p += 1 - uu; 

  }

  //p /= a_n * (a_h[0] * a_h[1]) * 16/9;
  p /= a_n * (a_h[0] * a_h[1]) * 4/3;
  
  return(p);
	
}


double kdeLoo_slow(double a_delta, int a_n, double *a_xx, double *a_vv, double a_h[2])
{
  
  double LOO = 0.0;
  
  for (int i=0; i<a_n; i++) 
  {
  
    double y = a_vv[i] - a_xx[i] - a_delta;
    double x = a_xx[i];
    
    if(y <= 0.0) continue;
    if(x <= 0.0) continue;

    x = log(x); y = log(y);
    
    for (int j=0; j<a_n; j++) 
    {
      
      if(i==j) continue;

      double yy = a_vv[j]-a_xx[j]-a_delta;

      if(yy <= 0.0) continue;
      if(a_xx[j] <= 0.0) continue;

      double z = (x-log(a_xx[j]))/a_h[0];
      double t = (y-log(yy))/a_h[1];
      //double zz = z*z;
      //double tt = t*t;
      double uu = z*z + t*t;

    //if((tt<1.0) && (tt>-1.0) && (zz<1.0) && (zz>-1.0)) LOO += (1-(t*t)) * (1-(z*z)); 
    if((uu<1.0) && (uu>-1.0)) LOO += 1 - uu; 

    }

  }
	
  //LOO/=a_n*(a_n-1)*(a_h[0]*a_h[1])*16/9;
  LOO/=a_n*(a_n-1)*(a_h[0]*a_h[1])*4/3;

  return(LOO);
  	
}

File: R/initFun.R
initQ<-function(fullData,xGrid,vGrid, deltaStar,tau=100,tol=1e-3,k=10L,N=100L,verbose=FALSE,h=NULL){
	
	storage.mode(N) <- "integer"
	
	obsData <- fullData[which(fullData$delta0*fullData$delta1==1),]
	obsData <- obsData[obsData$x!=0,]
	# obsData <- obsData[which(obsData$x+deltaStar<=obsData$v),]
	
	QcFit<-survfit(Surv(fullData$v-fullData$w,1-fullData$delta1)~1)
	Qc<-stepfun(QcFit$time,c(1,QcFit$surv),f=0)
	Snfit<-survfit(Surv(fullData$w,fullData$v,fullData$delta1)~1)
	Sn<-stepfun(Snfit$time,c(1,Snfit$surv),f=0)
	dGn<-1/Sn(fullData$w); 
	dGn[Sn(fullData$w)==0] <- 0
	dGn[dGn/sum(dGn)>100/nrow(fullData)] <- 0
	dGn<-dGn/sum(dGn)
	
	if(Qc(deltaStar)<=0)stop("Initialization NOT well defined.\a")
	
	w <- fullData$w
	
	delta <- deltaStar
	nnn <- as.integer(nrow(obsData))
	vv <- obsData$v; xx <- obsData$x
	storage.mode(vv) <- storage.mode(xx) <- "double"
	nx <- length(xGrid); nv <- length(vGrid)
	Q2 <- matrix(1.0,nrow=nv,ncol=nx)#*1.0
	storage.mode(tau) <- "double"
	
	#step <- tau/N
	
	W <- c(1,2^((1:(N-1))%%2+1),1)
	W <- W%*%t(W)
	storage.mode(W) <- "double"
	
	ab <- range(log(xx),na.rm=TRUE); storage.mode(ab) <- "double"
	cd <- range(log(vv-xx-delta),na.rm=TRUE); storage.mode(cd) <- "double"
	
	cat("a=",ab[1]," b=",ab[2],"\n")
	cat("c=",cd[1]," d=",cd[2],"\n")
	
	if(is.null(h)){
		
		hOpt <- rep(0.0,2)
		
		h1Grid <- seq(0.05,25,length.out=k)
		h2Grid <- seq(0.05,25,length.out=k)
		delta0 <- diff(h1Grid)[1]
		
		# h1Grid<-seq(max(1.5,bandwidth.nrd(xx-vv)-5),bandwidth.nrd(xx-vv)+5,by=0.5)
		# h2Grid<-seq(max(1.5,bandwidth.nrd(vv)-3),bandwidth.nrd(vv)+5,by=0.5)
		
		n1<-length(h1Grid); n2<-length(h2Grid)
		
		storage.mode(h1Grid) <- "double"
		storage.mode(h2Grid) <- "double"
		
		m<-1
		if(verbose)cat("\n")
		while(!2^m*delta0/(k^m)<tol){
			deltam <- diff(h1Grid)[1]
			z <- .C('initQ', a_delta=delta,a_ab=ab, a_cd=cd,a_n=nnn,a_xx=xx,a_vv=vv,a_h1=h1Grid,a_h2=h2Grid,a_n1=n1,a_n2=n2,a_N=N,a_xGrid=xGrid,a_vGrid=vGrid,a_nx=nx,a_nv=nv,a_Q2=Q2, a_hOpt= hOpt, a_Weights=W)
			hOpt <- z$a_hOpt
			if(verbose)cat("*")
			
			h1Grid <- seq(max(0.1,hOpt[1]-deltam-tol), hOpt[1]+deltam+tol,length.out=k)
			h2Grid <- seq(max(0.1,hOpt[2]-deltam-tol), hOpt[2]+deltam+tol,length.out=k)
			
			# cat(m," ")
			# print(range(h1Grid))
			# print(range(h2Grid))
			
			m <- m+1
			
			}
		if(verbose)cat("\n")
		
		}else{
			
			hOpt <- rep(0.0,2)
			
			h1Grid <- h[1]
			h2Grid <- h[2]
			n1<-1L; n2<-1L
			
			storage.mode(h1Grid) <- "double"
			storage.mode(h2Grid) <- "double"
			
			z <- .C('initQ', a_delta=delta,a_ab=ab, a_cd=cd,a_n=nnn,a_xx=xx,a_vv=vv,a_h1=h1Grid,a_h2=h2Grid,a_n1=n1,a_n2=n2,a_N=N,a_xGrid=xGrid,a_vGrid=vGrid,a_nx=nx,a_nv=nv,a_Q2=Q2, a_hOpt=hOpt, a_Weights=W)
			
			
			}
		
		
		Q2<-z$a_Q2
		hOpt <- z$a_hOpt
	
	if(verbose)cat("h Opt.")
	print(hOpt)
	attr(Q2,"hOpt") <- hOpt
	#Q2[1,] <- Q2[,1] <- 0.0
	
	normC <- -1.0
	z <- .C('Norm', a_normC=normC,a_xGrid=xGrid,a_vGrid=vGrid,a_nx=nx,a_nv=nv,a_Fun=Q2)
	Q2 <- Q2/z$a_normC
	
	
	return(list("Q1"=list("Sn"=Sn,"dGn"=dGn,"Qc"=Qc),"Q2"=Q2))
	
	}

initT<-function(fullData,Qc,dGn, deltaStar, xGrid,vGrid){

	w <- fullData$w
	dGn <- dGn[order(w)]
	w <- sort(w)
	n <- length(w)
	storage.mode(w) <- "double"
	
	nx <- length(xGrid); nv <- length(vGrid)
	
	x <- get("x",envir=environment(Qc))
	y <- get("y",envir=environment(Qc))
	m <- length(x)
	
	T1 <- matrix(0.0,nrow=nv,ncol=nx)
	T2 <- matrix(0.0,nrow=nv,ncol=nx)
	
	z <- .C('initT', a_T1=T1, a_T2=T2, a_Stimes=x, a_Sjumps=y, a_dGn=dGn,a_delta= deltaStar,a_n=n,a_m=m,a_ww=w,a_xGrid=xGrid,a_vGrid=vGrid,a_nx=nx,a_nv=nv)
	
	T1 <- z$a_T1
	T2 <- z$a_T2
	
	
	return(list("T1"=z$a_T1,"T2"=z$a_T2))
	
	}
